### PR TITLE
    [IMP] Merge technical group and debug mode , user with technical …

### PR DESCRIPTION
**Pull request by ajayjaviya**
Task : https://www.odoo.com/web#id=13763&view_type=form&model=project.task&menu_id=3942&action=327
Pad : https://pad.odoo.com/p/r.sD1wAuhQ2HSEnNkN

### CODE REVIEW

File: openerp/http.py [hunk 1]
```diff
@@ -315,7 +315,10 @@ def checked_call(___dbname, *a, **kw):
     def debug(self):
         """ Indicates whether the current request is in "debug" mode
         """
-        return 'debug' in self.httprequest.args
+        debug = 'debug' in self.httprequest.args
+        if not debug and self.httprequest.referrer:
```
Comment (tde-banana-odoo) [inline openerp/http.py]: What is the purpose of this change ? I mean, the use case that triggers this case ?

File: openerp/addons/base/tests/test_acl.py [hunk 2]
```diff
@@ -21,6 +21,9 @@ def setUp(self):
         self.tech_group = self.registry('ir.model.data').xmlid_to_object(self.cr, self.uid, GROUP_TECHNICAL_FEATURES)
         self.erp_manager_group = self.registry('ir.model.data').xmlid_to_object(self.cr, self.uid, 'base.group_erp_manager')
         self.erp_system_group = self.registry('ir.model.data').xmlid_to_object(self.cr, self.uid, 'base.group_system')
+        # As now we give technical rights to every user , so to adapt test case remove technical group from demo user
+        self.technical_group_id = self.registry('ir.model.data').xmlid_to_res_id(self.cr, self.uid, 'base.group_no_one')
```
Comment (tde-banana-odoo) [inline openerp/addons/base/tests/test_acl.py]: Is it necessary to add to to self ?

File: openerp/addons/base/res/res_users.py [hunk 3]
```diff
@@ -240,6 +240,8 @@ def _get_group(self,cr, uid, context=None):
             result.append(group_id)
             dummy,group_id = dataobj.get_object_reference(cr, SUPERUSER_ID, 'base', 'group_partner_manager')
             result.append(group_id)
+            dummy,group_id = dataobj.get_object_reference(cr, SUPERUSER_ID, 'base', 'group_no_one')
```
Comment (tde-banana-odoo) [inline openerp/addons/base/res/res_users.py]: Use xmlid_to_res_id (only for this one, do not change the previous one).

File: openerp/addons/base/res/res_users.py [hunk 4]
```diff
@@ -748,6 +750,9 @@ def update_user_groups_view(self, cr, uid, context=None):
                     app_name = app and app.name or _('Other')
                     xml2.append(E.separator(string=app_name, colspan="4", **attrs))
                     for g in gs:
+                        technical_group_id = self.pool['ir.model.data'].xmlid_to_res_id(cr, SUPERUSER_ID, 'base.group_no_one')
```
Comment (tde-banana-odoo) [inline openerp/addons/base/res/res_users.py]: Do the xmlid_tores_id only once + do a check once: is_technical = technical in groups .

Comment (tde-banana-odoo) [inline openerp/addons/base/res/res_users.py]: What is the purpose of this code ?

File: openerp/addons/base/ir/ir_ui_view.py [hunk 5]
```diff
@@ -662,7 +662,7 @@ def check_group(node):
             if node.get('groups'):
                 can_see = self.user_has_groups(
                     cr, user, groups=node.get('groups'), context=context)
-                if not can_see:
+                if not can_see or (request and not request.debug and 'base.group_no_one' in node.get('groups')):
```
Comment (tde-banana-odoo) [inline openerp/addons/base/ir/ir_ui_view.py]: Doesn't seem right; technical != bypassing groups validation. What are you trying to do here precisely ?

File: openerp/addons/base/ir/ir_ui_menu.py [hunk 6]
```diff
@@ -337,8 +338,8 @@ def load_menus_root(self, cr, uid, context=None):
 
 
     @api.cr_uid_context
-    @tools.ormcache_context('uid', keys=('lang',))
-    def load_menus(self, cr, uid, context=None):
+    @tools.ormcache_context('uid', 'debug', keys=('lang',))
+    def load_menus(self, cr, uid, debug=False, context=None):
```
Comment (tde-banana-odoo) [inline openerp/addons/base/ir/ir_ui_menu.py]: Why adding an unused parameter ? Moreover new parameters in critical methods like this should always be added at the end, for backward compatibility.

File: openerp/http.py (hunk 1 above)
Comment (aja-odoo) [inline openerp/http.py]: Issue is that debug mode is maintained only in request object if httprequest has debug args. so when field_view_get method call from JS Side (https://github.com/odoo-dev/odoo/blob/master-merge-technical-group-debug-mode-aja/addons/web/static/src/js/framework/model.js#L86), debug mode is not passed into httprequest args for that i have change in debug property of request object , check debug in httprquest.args and  if not found check in httprequest.referrer as falldack condition.

File: openerp/addons/base/tests/test_acl.py (hunk 2 above)
Comment (aja-odoo) [inline openerp/addons/base/tests/test_acl.py]: not necessary remove it.

File: openerp/addons/base/res/res_users.py (hunk 3 above)
Comment (aja-odoo) [inline openerp/addons/base/res/res_users.py]: done

File: openerp/addons/base/ir/ir_ui_menu.py (hunk 6 above)
Comment (aja-odoo) [inline openerp/addons/base/ir/ir_ui_menu.py]: load_menus should load from cache in both case debug is on/off , in both cases it must load relevant menu i mean if debug mode is off and menu have a technical group, menu should not visible untill user have a technical rights + debug mode is on. 
ok i add new parameter at the end.

File: openerp/addons/base/res/res_users.py (hunk 4 above)
Comment (aja-odoo) [inline openerp/addons/base/res/res_users.py]: if user have a technical rights , field should not visible untill debug mode is on.

File: openerp/addons/base/ir/ir_ui_view.py (hunk 5 above)
Comment (aja-odoo) [inline openerp/addons/base/ir/ir_ui_view.py]: can_see only check user has a group (base.group_no_one) , but in our case for technical rights we also need to check weather user is in debug mode or not.
